### PR TITLE
Simplifying the retrier

### PR DESCRIPTION
**Pull request by mariocynicys**
A proposal to simplify the retrier logic.

I am trying here to make the retrier manager responsible for every thing and the retriers only can change their statuses to direct the manager.

This is a bit simpler IMO, and **seems** to be race condition free, and I hope it is :).

### CODE REVIEW

File: watchtower-plugin/src/retrier.rs [hunk 1]
```diff
@@ -16,174 +16,248 @@ use crate::wt_client::WTClient;
...
+
+    fn set_status(&self, status: RetrierStatus) {
+        *self.status.lock().unwrap() = status;
+    }
+
+    pub fn is_running(&self) -> bool {
+        *self.status.lock().unwrap() == RetrierStatus::Running
+    }
+
+    pub fn ready(&self) -> bool {
```
Comment (sr-gi) [inline watchtower-plugin/src/retrier.rs]: nit: `needs_restart` maybe? It would be more descriptive. Ready is kind of ambiguous: ready for what?

File: watchtower-plugin/src/retrier.rs [hunk 2]
```diff
@@ -16,174 +16,248 @@ use crate::wt_client::WTClient;
...
+        // A retrier is ready to be re-started if it didn't fail and has some pending appointments.
+        // Such a retrier is one that has finished retrying all its pending appointments, but received some new
+        // appointments later on. It's safe to just re-start it instead of deleting it and instantiate a new one.
+        *self.status.lock().unwrap() == RetrierStatus::Success && self.has_pending_appointments()
+    }
+
+    pub fn retry(retrier: Self, max_elapsed_time_secs: u16, max_interval_time_secs: u16) {
```
Comment (sr-gi) [inline watchtower-plugin/src/retrier.rs]: Does it really make sense to have retry be a static method of the Retrier given a retrier is passed to it? It's a bit weird from a design perspective IMO.

File: watchtower-plugin/src/retrier.rs [hunk 3]
```diff
@@ -16,174 +16,248 @@ use crate::wt_client::WTClient;
...
+        let wt_client = retrier.wt_client.clone();
+        let tower_id = retrier.tower_id;
+        let max_elapsed_time = Some(Duration::from_secs(max_elapsed_time_secs as u64));
+        let max_interval = Duration::from_secs(max_interval_time_secs as u64);
+
```
Comment (sr-gi) [inline watchtower-plugin/src/retrier.rs]: All these temporary vars are not necessary given we are consuming the retrier.

File: watchtower-plugin/src/retrier.rs [hunk 4]
```diff
@@ -16,174 +16,248 @@ use crate::wt_client::WTClient;
...
             let wt_client = self.wt_client.lock().unwrap();
-            if wt_client.towers.get(&tower_id).is_none() {
+            if wt_client.towers.get(&self.tower_id).is_none() {
                 return Err(Error::permanent("Tower was abandoned. Skipping retry"));
             }
 
-            if self.pending_appointments.lock().unwrap().is_empty() {
+            if !self.has_pending_appointments() {
+                // will this ever happen ??
+                // FIXME: success/Ok() here instead so not to mark the tower as unreachable.
```
Comment (sr-gi) [inline watchtower-plugin/src/retrier.rs:240]: Yep, indeed. Currently, this would flag the tower as unreachable when it should not be the case. Either have a special Ok() return for this or just make it as if calls with no pending simply return Ok() and update the tests

File: watchtower-plugin/src/retrier.rs [hunk 5]
```diff
@@ -16,174 +16,248 @@ use crate::wt_client::WTClient;
...
+                        .wt_client
+                        .lock()
+                        .unwrap()
+                        .towers
+                        .contains_key(&tower_id)
+                    {
+                        log::info!("Skipping retrying abandoned tower {}", tower_id);
+                        continue;
+                    }
+                    self.add_pending_appointment(tower_id, locator);
```
Comment (sr-gi) [inline watchtower-plugin/src/retrier.rs:64]: This is a bit weird, isn't it. When a new retrier is created, this will just add it to retriers and wait for a new loop with a failed `recv` to start the retrier. Why not starting it here based on whether the retrier was just created or not? You could either return a bool and load the retrier based on that (or even a reference to the retrier, not sure how that would work wrt lifetimes)

Comment (sr-gi) [reply watchtower-plugin/src/retrier.rs:64]: You could also add an additional state to the `RetrierStatus` enum like `Ready` for freshly initialized retriers and check whether the retrier is running or ready. Currently the retriers start in `Success` mode, which is harmless but a bit misleading

File: watchtower-plugin/src/retrier.rs (hunk 4 above)
Comment (sr-gi) [reply watchtower-plugin/src/retrier.rs:240]: Giving this a second look I think it can, indeed, not happen.

`start` is called by `retry` which is only called if the `Retrier` is not running. Initially is called at least with one `locator`, and in the case one or more appointments were missed, it could be called with more than one.

`start` may be called over and over by the `ExponentialBackoff`, but only if some job was missed, so we are still ensuring at least one pending job.

All in all, I think this can only be called without actual data by a test, but not in the normal workflow of the retrier.

**Review by sr-gi — CHANGES_REQUESTED** [review]
I really like this revamp. Great job.

There are a few comments inline, but nothing major.

**Comment by sr-gi** [comment]
However, looks like the tests disagree with me 😢  https://github.com/sr-gi/rust-teos/actions/runs/3059572402/jobs/4937270925

File: watchtower-plugin/src/retrier.rs (hunk 1 above)
Comment (mariocynicys) [reply watchtower-plugin/src/retrier.rs]: I kept it like so to make it inclusive for the two cases:
1- when a fresh new retrier is ready to start.
2- when a finished retrier is ready to restart.
needs_restart accounts for case 2 only. agree there might be more descriptive naming, but I want it to kinda include both cases.

File: watchtower-plugin/src/retrier.rs (hunk 2 above)
Comment (mariocynicys) [reply watchtower-plugin/src/retrier.rs]: It wasn't static before, I try to remember why did I make it like that 😕.
Design-wise, I think it should be a normal method.

Comment (sr-gi) [reply watchtower-plugin/src/retrier.rs]: I think it's because of `tokio::spawn`, you cannot pass `&self`, therefore why you may have chosen to make it static.

Comment (sr-gi) [reply watchtower-plugin/src/retrier.rs]: That's, btw, why this was called by the `RetryManager` in my original approach and not by the `Retrier` itself 😉

**Comment by mariocynicys** [comment]
> However, looks like the tests disagree with me ￼ https://github.com/sr-gi/rust-teos/actions/runs/3059572402/jobs/4937270925

Oh, I think this is because the sleep we do when the channel is empty and we have done the housekeeping (removed old retriers and started new ones).

Will try to maybe add more sleeps in the e2e and see if that fixes it.

**Comment by sr-gi** [comment]
> > However, looks like the tests disagree with me ￼ https://github.com/sr-gi/rust-teos/actions/runs/3059572402/jobs/4937270925
> 
> Oh, I think this is because the sleep we do when the channel is empty and we have done the housekeeping (removed old retriers and started new ones).
> 
> Will try to maybe add more sleeps in the e2e and see if that fixes it.

The issue with that kind of errors is that they may be hard to reproduce (they never get trigger in my local setup for instance), therefore hard to know if they were doe to a non-sufficient sleep or an actual race condition. #84 was actually failing exactly there 😅

File: watchtower-plugin/src/retrier.rs (hunk 5 above)
Comment (mariocynicys) [reply watchtower-plugin/src/retrier.rs:64]: > This is a bit weird, isn't it. When a new retrier is created, this will just add it to retriers and wait for a new loop with a failed recv to start the retrier. Why not starting it here based on whether the retrier was just created or not? You could either return a bool and load the retrier based on that (or even a reference to the retrier, not sure how that would work wrt lifetimes)

~~Ummm, I think this will actually fix the failing e2e tests, by starting the retriers right away.~~ (edit: this actually isn't correct, the draining of the channel happens very quickly already, since when the channel is loaded, the retrier doesn't do the 1s wait)
But this kinda looks counter intuitive for me :(.

Comment (mariocynicys) [reply watchtower-plugin/src/retrier.rs:64]: > You could also add an additional state to the RetrierStatus enum like Ready for freshly initialized retriers and check whether the retrier is running or ready. Currently the retriers start in Success mode, which is harmless but a bit misleading

Also one other solution would to rename success and failure to more generic terms (like good/bad). This would clear the confusion if we state what these terms mean (e.g. good = this retrier can be used, bad = get rid of me right away!!).

File: watchtower-plugin/src/retrier.rs (hunk 2 above)
Comment (mariocynicys) [reply watchtower-plugin/src/retrier.rs]: > That's, btw, why this was called by the RetryManager in my original approach and not by the Retrier itself ￼

+1, but it makes more sense in the retrier now cuz it changes the retrier status. which the retrier itself should only be responsible of.

Comment (sr-gi) [reply watchtower-plugin/src/retrier.rs]: Yeah, it also makes it so the `retriers` map can store `Retrier` instead of `Arc<Mutex<Retrier>>`

File: watchtower-plugin/src/retrier.rs (hunk 1 above)
Comment (sr-gi) [reply watchtower-plugin/src/retrier.rs]: Fair enough, but then, at the very least, the docs need to be updated. Right now they only read about re-starting.

Notice this only works because you are overloading `RetrierStatus::Success` for just created and completed retriers, if you end up adding a new variant to the enum it'll need to be updated

**Comment by sr-gi** [comment]
@meryacine In any case, lmk if there's any way I can help you test/debug the error so we can speed up this

File: watchtower-plugin/src/retrier.rs [hunk 6]
```diff
@@ -64,18 +64,22 @@ impl RetryManager {
...
                         }
                     }
-                    // Sleep to not waste a lot of CPU cycles.
-                    tokio::time::sleep(Duration::from_secs(1)).await;
+                    // When max elapsed time is zero, this is probably a test environment and we want
+                    // things to be really fast. So don't sleep.
+                    if self.max_elapsed_time_secs != 0 {
+                        // Sleep to not waste a lot of CPU cycles.
+                        tokio::time::sleep(Duration::from_secs(1)).await;
+                    }
```
Comment (mariocynicys) [inline watchtower-plugin/src/retrier.rs]: This part should let the test pass. I hope.
But the retrier thread will do a lot of busy waiting when ran by tests, which isn't ideal. The best idea I have is to make this one second sleep a variable that we can control from tests and account for it (e.g. in the tests we set it to 1 and sleep 1 sec to account for it).

Comment (mariocynicys) [reply watchtower-plugin/src/retrier.rs]: Didn't fix it: https://github.com/meryacine/rust-teos/actions/runs/3066853645/jobs/4952503002#step:9:6925

Comment (sr-gi) [reply watchtower-plugin/src/retrier.rs]: So if I'm not mistaken the issue here is that the retrier is not flagging the tower as unreachable fast enough, is it not? Basically we are checking the state before the retrier has given up. Given the retrier is set to give up "immediately" (`"watchtower-max-retry-time": 0`) we can just add a small sleep in the test after sending the payment.

Comment (mariocynicys) [reply watchtower-plugin/src/retrier.rs]: Yeah a small sleep should fix it, I was trying to avoid it though.
I will try to make that one second sleep settable from the command line so tests doesn't depend on other code.

Comment (sr-gi) [reply watchtower-plugin/src/retrier.rs]: Even though it's not elegant, it may still be necessary. Otherwise, the test may randomly fail. It's basically a race between the retrier failing and the test checking the tower state.

**Comment by mariocynicys** [comment]
@sr-gi any clue why this test doesn't stop? I only added a 1 sec sleep :/
https://github.com/meryacine/rust-teos/actions/runs/3067235568/jobs/4953309158

File: watchtower-plugin/src/retrier.rs [hunk 7]
```diff
@@ -16,174 +16,248 @@ use crate::wt_client::WTClient;
...
-                            }
+                Err(TryRecvError::Empty) => {
+                    // Keep only running retriers and retriers ready to be started/re-started.
+                    // This will remove failed ones and ones finished successfully and have no pending appointments.
+                    //
+                    // Note that a failed retrier could have received some new appointments to retry. In this case, we don't try to send
+                    // them because we know that that tower is unreachable. We most likely received these new appointments because the tower
+                    // was marked as temporary unreachable during the retrial process so the sender didn't know that the tower is unreachable.
+                    // But this can also happen if the user manually retries the tower just right after (within 1 second time frame) it being
+                    // marked as unreachable. The user needs to wait some time and try again when we have remove the retrier for that tower.
```
Comment (sr-gi) [inline watchtower-plugin/src/retrier.rs]: I think this issue can be fixed by flagging the tower as unreachable here, instead of as a result of the backoff strategy, that is:

The retrier is flagged as failed by the backoff failure, then it gets here, its drained as failed and the corresponding tower is flagged. In that way the user won't be able to manually retry until the retrier is actually wiped, given the tower being unreachable is a requirement for that.

**Review by sr-gi — CHANGES_REQUESTED** [review]
Just the single inline comment.

LGTM otherwise. It'll need squashing.

File: watchtower-plugin/src/retrier.rs [hunk 8]
```diff
@@ -16,174 +16,242 @@ use crate::wt_client::WTClient;
...
-                            } else {
-                                log::info!("Skipping retrying abandoned tower {}", tower_id);
-                            }
+                Err(TryRecvError::Empty) => {
+                    // Keep only running retriers and retriers ready to be started/re-started.
+                    // This will remove failed ones and ones finished successfully and have no pending appointments.
+                    //
+                    // Note that a failed retrier could have received some new appointments to retry. In this case, we don't try to send
+                    // them because we know that that tower is unreachable. We most likely received these new appointments because the tower
+                    // was marked as temporary unreachable during the retrial process so the sender didn't know that the tower is unreachable.
```
Comment (sr-gi) [inline watchtower-plugin/src/retrier.rs]: I'd reword this to: "We most likely received these new appointments while the tower was still flagged as temporarily unreachable when cleaning up after giving up retrying."

File: watchtower-plugin/src/retrier.rs [hunk 9]
```diff
@@ -265,6 +337,28 @@ impl Retrier {
...
     }
+
+    /// For failed retriers, we set the tower to unreachable at drop time and not after backoff is complete.
+    /// Because if we set the tower as unreachable after backoff is complete, the user might perform manual retry
+    /// for that tower but this action will never be performed nor that we give error to the user.
+    /// This is because a failed retrier is never re-started, and is removed immediately by the retry manager
+    /// when it wakes from sleep (even if that failed retrier had some pending appointments).
+    ///
+    /// Setting tower status as unreachable at drop time keeps the tower in temporary unreachable state
+    /// in any other time, so it gives back an error when it is manually retried by the user.
```
Comment (sr-gi) [inline watchtower-plugin/src/retrier.rs]: I think this is way too descriptive.

The docs should read a 1 / 2 lines describing what the method does (setting the status to unreachable) and then having a longer description is needed after a two lines break, that way the longer description is only shown if the method is expanded on the docs.

With respect to the longer description, I think you don't need to motivate it that much, it has already been mentioned twice elsewhere in the codebase why the status is not set there (which we are kind off inclined at the moment given we're moving that piece of code, but it may not be that noticeable for new contributors).

Just saying this is supposed to be done after the retrier is dropped so the issue with the user trying to retry is avoided should be enough.

**Review by sr-gi — APPROVED** [review]
Two minor comments regarding the docs.

LGTM otherwise. Needs squashing.

File: watchtower-plugin/src/retrier.rs (hunk 9 above)
Comment (mariocynicys) [reply watchtower-plugin/src/retrier.rs]: I am sometimes extra descriptive cuz my memory suck :disappointed:. Will fix that now.
